File: frontend/src/types/organization-profile.ts
/** Reviewed tenant profile DTO. Never includes settings, branding or domains. */
export interface OrganizationContact { purpose: string; email: string }
export type OrganizationContactPurpose = 'security' | 'privacy' | 'billing' | 'support';
export type OrganizationProfileLanguage = 'en' | 'de' | 'fr';
export interface OrganizationProfile {
  id: string;
  name: string;
  slug: string;
  legal_name: string;
  industry: string;
  country_code: string;
  timezone: string;
  default_language: string;
  supported_languages: string[];
  employee_count_range: string;
  status: string;
  tier: string;
  fiscal_year_start_month: number;
  fiscal_year_start_day: number;
  contacts: OrganizationContact[];
  version: number;
  updated_at: string;
}

/** ABAC may omit fields or mask values. This is not an editable full DTO. */
export type OrganizationProfileProjection = Partial<OrganizationProfile>;
export interface OrganizationProfileMeta {
  schema_version: 1;
  scope: 'tenant_organization_profile';
  editable: boolean;
}
export interface OrganizationProfileEnvelope {
  data: OrganizationProfileProjection;
  /** Missing/invalid metadata fails closed for editing. */
  meta?: OrganizationProfileMeta;
}
export type OrganizationProfileEditable = Pick<OrganizationProfile,
  'name' | 'legal_name' | 'industry' | 'country_code' | 'timezone' | 'default_language' |
  'supported_languages' | 'employee_count_range' | 'fiscal_year_start_month' | 'fiscal_year_start_day' | 'contacts'>;
export interface OrganizationProfileUpdateInput extends Partial<OrganizationProfileEditable> {
  expected_version: number;
  reason: string;
}

File: internal/models/organization_profile.go
package models

import (
	"errors"
	"time"
)

const OrganizationProfileMaximumVersion int64 = (1 << 53) - 1
const OrganizationProfileScope = "tenant_organization_profile"

var (
	ErrOrganizationProfileScope       = errors.New("organization profile tenant or actor context is invalid")
	ErrOrganizationProfileNotFound    = errors.New("organization profile is unavailable for this tenant")
	ErrOrganizationProfileConflict    = errors.New("organization profile changed; reload before saving")
	ErrOrganizationProfileUnavailable = errors.New("organization profile is temporarily unavailable")
)

// OrganizationContact is informational, not an email transport configuration,
// billing instruction, verified domain, or grant of administrator authority.
type OrganizationContact struct {
	Purpose string `json:"purpose"`
	Email   string `json:"email"`
}

// OrganizationProfile projects only reviewed profile fields. It never embeds
// raw settings, branding, metadata, credentials, or parent-tenant references.
type OrganizationProfile struct {
	ID                   string                `json:"id"`
	Name                 string                `json:"name"`
	Slug                 string                `json:"slug"`
	LegalName            string                `json:"legal_name"`
	Industry             string                `json:"industry"`
	CountryCode          string                `json:"country_code"`
	Timezone             string                `json:"timezone"`
	DefaultLanguage      string                `json:"default_language"`
	SupportedLanguages   []string              `json:"supported_languages"`
	EmployeeCountRange   string                `json:"employee_count_range"`
	Status               string                `json:"status"`
	Tier                 string                `json:"tier"`
	FiscalYearStartMonth int                   `json:"fiscal_year_start_month"`
	FiscalYearStartDay   int                   `json:"fiscal_year_start_day"`
	Contacts             []OrganizationContact `json:"contacts"`
	Version              int64                 `json:"version"`
	UpdatedAt            time.Time             `json:"updated_at"`
}

// Editable proves that this projection was not masked/hidden. It is not a
// settings:configure grant: the mutation policy is evaluated independently.
type OrganizationProfileMeta struct {
	SchemaVersion int    `json:"schema_version"`
	Scope         string `json:"scope"`
	Editable      bool   `json:"editable"`
}

type OrganizationProfileResponse struct {
	Data *OrganizationProfile    `json:"data"`
	Meta OrganizationProfileMeta `json:"meta"`
}

// OrganizationProfileUpdateInput uses presence-aware fields for the legacy PUT
// URL: absent fields are preserved, explicit null is rejected by the handler.
// Read-only identifiers, lifecycle, tier, raw settings and domains are not inputs.
type OrganizationProfileUpdateInput struct {
	ExpectedVersion      int64                  `json:"expected_version"`
	Reason               string                 `json:"reason"`
	Name                 *string                `json:"name,omitempty"`
	LegalName            *string                `json:"legal_name,omitempty"`
	Industry             *string                `json:"industry,omitempty"`
	CountryCode          *string                `json:"country_code,omitempty"`
	Timezone             *string                `json:"timezone,omitempty"`
	DefaultLanguage      *string                `json:"default_language,omitempty"`
	SupportedLanguages   *[]string              `json:"supported_languages,omitempty"`
	EmployeeCountRange   *string                `json:"employee_count_range,omitempty"`
	FiscalYearStartMonth *int                   `json:"fiscal_year_start_month,omitempty"`
	FiscalYearStartDay   *int                   `json:"fiscal_year_start_day,omitempty"`
	Contacts             *[]OrganizationContact `json:"contacts,omitempty"`
}

// OrganizationProfileValidationError exposes only reviewed field/code labels,
// never the rejected value, mail parser detail, or a provider/database error.
type OrganizationProfileValidationError struct {
	Field string `json:"field"`
	Code  string `json:"code"`
}

func (e *OrganizationProfileValidationError) Error() string {
	return "organization profile field is invalid"
}
